services/subnet_manager/src/main/java/com/futurewei/alcor/subnet/controller/SubnetController.java

```diff
@@ -0,0 +1,251 @@
+import com.futurewei.alcor.subnet.service.SubnetService;
+import com.futurewei.alcor.subnet.utils.RestPreconditionsUtil;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static org.springframework.web.bind.annotation.RequestMethod.*;
+
+@RestController
+public class SubnetController {
+
+    @Autowired
+    private SubnetDatabaseService subnetDatabaseService;
+
+    @Autowired
+    private SubnetService subnetService;
+
+    @RequestMapping(
+            method = GET,
+            value = {"/project/{projectid}/subnets/{subnetId}", "v4/{projectid}/subnets/{subnetId}"})
+    public SubnetStateJson getSubnetStateById(@PathVariable String projectid, @PathVariable String subnetId) throws Exception {
+
+        SubnetState subnetState = null;
+
+        try {
+            RestPreconditionsUtil.verifyParameterNotNullorEmpty(projectid);
+            RestPreconditionsUtil.verifyParameterNotNullorEmpty(subnetId);
+            RestPreconditionsUtil.verifyResourceFound(projectid);
+
+            subnetState = this.subnetDatabaseService.getBySubnetId(subnetId);
+        } catch (ParameterNullOrEmptyException e) {
+            //TODO: REST error code
+            throw new Exception(e);
+        }
+
+        if (subnetState == null) {
+            //TODO: REST error code
+            return new SubnetStateJson();
+        }
+
+        return new SubnetStateJson(subnetState);
+    }
+
+    @RequestMapping(
+            method = POST,
+            value = {"/project/{projectid}/subnets", "v4/{projectid}/subnets"})
+    @ResponseStatus(HttpStatus.CREATED)
+    public SubnetStateJson createSubnetState(@PathVariable String projectid, @RequestBody SubnetStateJson resource) throws Exception {
+        SubnetState subnetState = null;
+
+        try {
+            RestPreconditionsUtil.verifyParameterNotNullorEmpty(projectid);
+            RestPreconditionsUtil.verifyResourceNotNull(resource.getSubnet());
+
+            // TODO: Create a verification framework for all resources
+            SubnetState inSubnetState = resource.getSubnet();
+            RestPreconditionsUtil.verifyResourceFound(inSubnetState.getVpcId());
+            RestPreconditionsUtil.populateResourceProjectId(inSubnetState, projectid);
+
+            this.subnetDatabaseService.addSubnet(inSubnetState);
+
+            subnetState = this.subnetDatabaseService.getBySubnetId(inSubnetState.getId());
+            if (subnetState == null) {
+                throw new ResourcePersistenceException();
+            }
+
+            // Verify VPC ID
+            VpcStateJson vpcResponse = this.subnetService.verifyVpcId(projectid, inSubnetState.getVpcId());
+            if (vpcResponse == null) {
+                throw new ResourcePersistenceException();
```

Comment:
Why vcResponse = ResourcePersistence?

Reply:
You might want to throw a more specific exception for the case of "vpc doesn't exist"?

Reply:
At the beginning, I set it to a specific exception because UTs are convenient for testing. Now I make it throw new Exception()

Reply:
a new generic Exception? I don't think you want to do that as you can't tell which block of codes threw the exception.